reactos/drivers/usb/usbhub_new/usbhub.c

```diff
@@ -0,0 +1,5143 @@
+    Urb->Hdr.Length = sizeof(struct _URB_CONTROL_DESCRIPTOR_REQUEST);
+
+    Urb->TransferBufferLength = NumberOfBytes;
+    Urb->TransferBuffer = ConfigDescriptor;
+    Urb->DescriptorType = USB_CONFIGURATION_DESCRIPTOR_TYPE;
+
+    if (DeviceExtension->ExtensionType == USBH_EXTENSION_TYPE_HUB ||
+        DeviceExtension->ExtensionType == USBH_EXTENSION_TYPE_PARENT)
+    {
+        Status = USBH_FdoSyncSubmitUrb(DeviceObject, (PURB)Urb);
+    }
+    else
+    {
+        Status = USBH_SyncSubmitUrb(DeviceObject, (PURB)Urb);
+    }
+
+    if (OutLength)
+    {
+        *OutLength = Urb->TransferBufferLength;
+    }
+
+    if (Urb)
+    {
+        ExFreePool(Urb);
+    }
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_GetConfigurationDescriptor(IN PDEVICE_OBJECT DeviceObject,
+                                IN PUSB_CONFIGURATION_DESCRIPTOR * pConfigurationDescriptor)
+{
+    PUSB_CONFIGURATION_DESCRIPTOR ConfigDescriptor;
+    NTSTATUS Status;
+    ULONG Length;
+
+    DPRINT("USBH_GetConfigurationDescriptor: ... \n");
+
+    ConfigDescriptor = ExAllocatePoolWithTag(NonPagedPool, 0xFF, USB_HUB_TAG);
+
+    if (!ConfigDescriptor)
+    {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto ErrorExit;
+    }
+
+    RtlZeroMemory(ConfigDescriptor, 0xFF);
+
+    Status = USBH_SyncGetDeviceConfigurationDescriptor(DeviceObject,
+                                                       ConfigDescriptor,
+                                                       0xFF,
+                                                       &Length);
+
+    if (!NT_SUCCESS(Status))
+    {
+        goto ErrorExit;
+    }
+
+    if (Length < sizeof(USB_CONFIGURATION_DESCRIPTOR))
+    {
+        Status = STATUS_DEVICE_DATA_ERROR;
+        goto ErrorExit;
+    }
+
+    if (Length < ConfigDescriptor->wTotalLength)
+    {
+        Status = STATUS_DEVICE_DATA_ERROR;
+        goto ErrorExit;
+    }
+
+    if (ConfigDescriptor->wTotalLength > 0xFF)
+    {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto ErrorExit;
+    }
+
+    *pConfigurationDescriptor = ConfigDescriptor;
+
+     return Status;
+
+ErrorExit:
+
+    if (ConfigDescriptor)
+    {
+        ExFreePool(ConfigDescriptor);
+    }
+
+    *pConfigurationDescriptor = NULL;
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_SyncGetHubDescriptor(IN PUSBHUB_FDO_EXTENSION HubExtension)
+{
+    PUSB_EXTHUB_INFORMATION_0 ExtendedHubInfo;
+    ULONG NumberPorts;
+    PUSBHUB_PORT_DATA PortData;
+    USHORT RequestValue;
+    ULONG NumberOfBytes;
+    NTSTATUS Status;
+    PUSB_HUB_DESCRIPTOR HubDescriptor;
+    ULONG ix;
+
+    DPRINT("USBH_SyncGetHubDescriptor: ... \n");
+
+    ExtendedHubInfo = ExAllocatePoolWithTag(NonPagedPool,
+                                            sizeof(USB_EXTHUB_INFORMATION_0),
+                                            USB_HUB_TAG);
+
+    if (!ExtendedHubInfo)
+    {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto ErrorExit;
```

Comment:
This code path will use HubDescriptor uninitialized
error C4703: potentially uninitialized local pointer variable 'HubDescriptor' used